### PR TITLE
[maskedtensor] use masked_softmax for forward/backward instead of regular softmax

### PR DESCRIPTION
**Pull request by george-qi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85845

### CODE REVIEW

File: torch/masked/maskedtensor/_ops_refs.py [hunk 1]
```diff
@@ -313,8 +313,7 @@ def _softmax(func, *args, **kwargs):
     _check_args_kwargs_length(args, kwargs, f"__torch_dispatch__, {func}", len_args=3, len_kwargs=0)
     data = _get_data(args[0])
     mask = _maybe_get_mask(args[0])
-    input_data = data.masked_fill(~mask, float("-inf"))
-    result_data = func(input_data, args[1], args[2])
+    result_data = torch.ops.aten._masked_softmax(data, ~mask, args[1], 1)
```
Comment (george-qi) [inline torch/masked/maskedtensor/_ops_refs.py:316]: 1 is necessary because of https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L143

**Comment by george-qi** [comment]
Using a similar benchmark as https://github.com/pytorch/pytorch/pull/76996#issuecomment-1123306850:

In general, when comparing just the native kernels we notice that speedups of around 1.5-2.25x across all levels of sparsity (though not in every single example), especially for slightly larger sizes. 

```
import torch
from torch.masked import masked_tensor
import time
import numpy as np
import pandas as pd
import itertools


def benchmark_torch_function(iters, f, *args, **kwargs):
    f(*args, **kwargs)
    if torch.cuda.is_available():
        torch.cuda.synchronize()
        start_event = torch.cuda.Event(enable_timing=True)
        end_event = torch.cuda.Event(enable_timing=True)
        start_event.record()
    else:
        t0 = time.time()
    for i in range(iters):
        f(*args, **kwargs)
    if torch.cuda.is_available():
        end_event.record()
        torch.cuda.synchronize()
        return start_event.elapsed_time(end_event)
    else:
        return (time.time() - t0)

def fn1(mt, dim):
    data = mt.get_data()
    mask = mt.get_mask()

    input_data = data.masked_fill(~mask, float("-inf"))
    result_data = torch.ops.aten._softmax(input_data, dim, False)

def fn2(mt, dim):
    data = mt.get_data()
    mask = mt.get_mask()

    result_data = torch.ops.aten._masked_softmax(data, ~mask, dim, 1)

def fn3(grad, output, dim, dtype):
    grad_data = grad.get_data()
    output_data = output.get_data()
    mask = grad.get_mask()
    grad_data = grad_data.masked_fill(~mask, 1)
    output_data = output_data.masked_fill(~mask, 0)
    result_data = torch.ops.aten._softmax_backward_data(grad_data, output_data, dim, dtype)

def fn4(grad, output, mask, dim):
    grad_data = grad.get_data()
    output_data = output.get_data()
    mask = grad.get_mask()
    result_data = torch.ops.aten._masked_softmax_backward(grad_data, output_data, ~mask, dim)


df_lst = []
iters = 100
sizes = [128, 512]

for is_cpu in [True, False]:
    for dtype in [torch.float32]:
        for (size1, size2) in itertools.product(sizes, sizes):
            for mask_perc in [0.01, 0.1, 0.9, 0.99]:
                t = torch.rand((size1, size2), dtype=dtype)
                mask = t > mask_perc
                dim = t.dim() - 1

                output_data = torch.ops.aten._masked_softmax(t, ~mask, dim, 1)
                grad_init = torch.randn((size1, size2))

                if not is_cpu:
                    t = t.cuda()
                    mask = mask.cuda()
                    output_data = output_data.cuda()
                    grad_init = grad_init.cuda()

                mt = masked_tensor(t, mask)
                output = masked_tensor(output_data, mask)
                grad = masked_tensor(grad_init, mask)

                f1_times, f2_times, f3_times, f4_times = [], [], [], []

                for _ in range(10):
                    f1_times.append(benchmark_torch_function(iters, fn1, mt, dim) / iters)
                    f2_times.append(benchmark_torch_function(iters, fn2, mt, dim) / iters)
                    f3_times.append(benchmark_torch_function(iters, fn3, grad, output, dim, dtype) / iters)
                    f4_times.append(benchmark_torch_function(iters, fn4, grad, output, mask, dim) / iters)

                df_lst.append([
                    {
                        'is_cpu': is_cpu,
                        'size': (size1, size2),
                        'masked_out': mask_perc,
                        'dtype': dtype,
                        'f1_mean': np.mean(f1_times),
                        'f2_mean': np.mean(f2_times),
                        'f1_std': np.std(f1_times),
                        'f2_std': np.std(f2_times),
                        'f3_mean': np.mean(f3_times),
                        'f4_mean': np.mean(f4_times),
                        'f3_std': np.std(f3_times),
                        'f4_std': np.std(f4_times),
                    }
                ])
                
                print (f"size: ({size1}, {size2}), % masked out: {mask_perc}, cpu: {is_cpu}, dtype: {dtype}")
                print (f"f1. mean: {f1_mean}, std: {f1_std}")
                print (f"f2. mean: {f2_mean}, std: {f2_std}")
                print (f"f3. mean: {f3_mean}, std: {f3_std}")
                print (f"f4. mean: {f4_mean}, std: {f4_std}")
                print ()

df = pd.DataFrame(df_lst)

print(df.to_string())


```

`dtype = torch.float32`: 
| is_cpu | size | masked out % | softmax | masked softmax | masked softmax speedup | softmax backwards | masked softmax backwards | masked softmax backwards speedup
|-------:|-----:|---------------:|--------:|-----------------:|---------------------------:|--------------------:|-----------------------------:|---------------------------------------:|
| True | (128, 128) | 0.01 | 0.1399 ± 0.0082 | 0.3120 ± 0.0314 | 0.45 | 0.2072 ± 0.0030 | 0.3313 ± 0.0261 | 0.63
 | True | (128, 128) | 0.1 | 0.1547 ± 0.0045 | 0.3506 ± 0.0049 | 0.44 | 0.2448 ± 0.0097 | 0.3628 ± 0.0019 | 0.67
 | True | (128, 128) | 0.9 | 0.1592 ± 0.0020 | 0.3442 ± 0.0041 | 0.46 | 0.2508 ± 0.0040 | 0.3620 ± 0.0044 | 0.69
 | True | (128, 128) | 0.99 | 0.1453 ± 0.0024 | 0.3460 ± 0.0192 | 0.42 | 0.2317 ± 0.0198 | 0.3595 ± 0.0062 | 0.64
 | True | (128, 512) | 0.01 | 0.4496 ± 0.0365 | 0.3821 ± 0.0060 | 1.18 | 0.5747 ± 0.0112 | 0.4006 ± 0.0060 | 1.43
 | True | (128, 512) | 0.1 | 0.4600 ± 0.0119 | 0.3798 ± 0.0039 | 1.21 | 0.6308 ± 0.0094 | 0.4135 ± 0.0412 | 1.53
 | True | (128, 512) | 0.9 | 0.4598 ± 0.0061 | 0.3741 ± 0.0038 | 1.23 | 0.6311 ± 0.0062 | 0.3970 ± 0.0056 | 1.59
 | True | (128, 512) | 0.99 | 0.4373 ± 0.0051 | 0.3762 ± 0.0239 | 1.16 | 0.5906 ± 0.0079 | 0.3897 ± 0.0020 | 1.52
 | True | (512, 128) | 0.01 | 0.4290 ± 0.0061 | 0.3941 ± 0.0072 | 1.09 | 0.5974 ± 0.0556 | 0.4167 ± 0.0132 | 1.43
 | True | (512, 128) | 0.1 | 0.4874 ± 0.0247 | 0.4150 ± 0.0306 | 1.17 | 0.7170 ± 0.0982 | 0.4298 ± 0.0089 | 1.67
 | True | (512, 128) | 0.9 | 0.4862 ± 0.0147 | 0.4086 ± 0.0103 | 1.19 | 0.7690 ± 0.0697 | 0.4285 ± 0.0067 | 1.79
 | True | (512, 128) | 0.99 | 0.4644 ± 0.0133 | 0.4074 ± 0.0123 | 1.14 | 0.6816 ± 0.0629 | 0.4200 ± 0.0124 | 1.62
 | True | (512, 512) | 0.01 | 0.7017 ± 0.0202 | 0.4833 ± 0.0091 | 1.45 | 0.9865 ± 0.0192 | 0.4851 ± 0.0108 | 2.03
 | True | (512, 512) | 0.1 | 0.7226 ± 0.0181 | 0.4910 ± 0.0199 | 1.47 | 1.0949 ± 0.1076 | 0.4867 ± 0.0241 | 2.25
 | True | (512, 512) | 0.9 | 0.7562 ± 0.0490 | 0.4842 ± 0.0450 | 1.56 | 1.0464 ± 0.0270 | 0.4826 ± 0.0095 | 2.17
 | True | (512, 512) | 0.99 | 0.6970 ± 0.0137 | 0.4517 ± 0.0088 | 1.54 | 1.0021 ± 0.0358 | 0.4699 ± 0.0102 | 2.13
 | False | (128, 128) | 0.01 | 0.0663 ± 0.0022 | 0.0443 ± 0.0010 | 1.50 | 0.1232 ± 0.0033 | 0.0714 ± 0.0014 | 1.73
 | False | (128, 128) | 0.1 | 0.0650 ± 0.0012 | 0.0437 ± 0.0009 | 1.49 | 0.1203 ± 0.0020 | 0.0710 ± 0.0013 | 1.69
 | False | (128, 128) | 0.9 | 0.0659 ± 0.0014 | 0.0440 ± 0.0012 | 1.50 | 0.1215 ± 0.0026 | 0.0712 ± 0.0019 | 1.71
 | False | (128, 128) | 0.99 | 0.0658 ± 0.0018 | 0.0438 ± 0.0008 | 1.50 | 0.1213 ± 0.0028 | 0.0717 ± 0.0013 | 1.69
 | False | (128, 512) | 0.01 | 0.0657 ± 0.0015 | 0.1564 ± 0.3386 | 0.42 | 0.1211 ± 0.0010 | 0.0716 ± 0.0012 | 1.69
 | False | (128, 512) | 0.1 | 0.0651 ± 0.0011 | 0.0430 ± 0.0007 | 1.51 | 0.1197 ± 0.0009 | 0.0708 ± 0.0011 | 1.69
 | False | (128, 512) | 0.9 | 0.0659 ± 0.0020 | 0.0437 ± 0.0012 | 1.51 | 0.1210 ± 0.0018 | 0.0706 ± 0.0006 | 1.71
 | False | (128, 512) | 0.99 | 0.0650 ± 0.0009 | 0.0438 ± 0.0014 | 1.48 | 0.1197 ± 0.0020 | 0.0705 ± 0.0010 | 1.70
 | False | (512, 128) | 0.01 | 0.0649 ± 0.0007 | 0.0436 ± 0.0008 | 1.49 | 0.1187 ± 0.0016 | 0.0706 ± 0.0009 | 1.68
 | False | (512, 128) | 0.1 | 0.0656 ± 0.0015 | 0.0441 ± 0.0010 | 1.49 | 0.1206 ± 0.0017 | 0.0710 ± 0.0010 | 1.70
 | False | (512, 128) | 0.9 | 0.0648 ± 0.0014 | 0.0438 ± 0.0011 | 1.48 | 0.1192 ± 0.0017 | 0.0705 ± 0.0015 | 1.69
 | False | (512, 128) | 0.99 | 0.0651 ± 0.0009 | 0.0432 ± 0.0011 | 1.51 | 0.1204 ± 0.0023 | 0.0718 ± 0.0017 | 1.68
 | False | (512, 512) | 0.01 | 0.0652 ± 0.0019 | 0.0438 ± 0.0009 | 1.49 | 0.1206 ± 0.0016 | 0.0727 ± 0.0024 | 1.66
 | False | (512, 512) | 0.1 | 0.0645 ± 0.0013 | 0.0437 ± 0.0009 | 1.48 | 0.1192 ± 0.0018 | 0.0722 ± 0.0030 | 1.65
 | False | (512, 512) | 0.9 | 0.0642 ± 0.0014 | 0.0434 ± 0.0010 | 1.48 | 0.1191 ± 0.0014 | 0.0713 ± 0.0018 | 1.67
 | False | (512, 512) | 0.99 | 0.0649 ± 0.0022 | 0.0437 ± 0.0007 | 1.48 | 0.1194 ± 0.0014 | 0.0718 ± 0.0024 | 1.66

File: torch/masked/maskedtensor/core.py [hunk 2]
```diff
@@ -51,7 +51,7 @@ def _tensors_match(a, b, exact=True):
         )
     if exact:
         return (a.dim() == b.dim()) and torch.eq(a, b).all().item()
-    return (a.dim() == b.dim()) and torch.allclose(a, b)
+    return (a.dim() == b.dim()) and torch.allclose(a, b, atol=1e-4)
```
Comment (george-qi) [inline torch/masked/maskedtensor/core.py:54]: was getting some errors on the order of 1e-7, this is a buffer just in case

Comment (george-qi) [inline torch/masked/maskedtensor/core.py:54]: was getting some errors on the order of 1e-7, this is a buffer just in case

File: torch/masked/maskedtensor/_ops_refs.py (hunk 1 above)
Comment (cpuhrsch) [reply torch/masked/maskedtensor/_ops_refs.py:316]: Based on the comments this maps to 
```
// If input is [B, H, T, T] and mask is [B, T]
  // we have special fast kernel
  // mask_type == 1 => mask_ is a src_key_padding_mask
```

But mask here should be of shape `[B, H, T, T]`. 

Based on the comment you're referencing It seems like the kernel on CPU checks the the mask is of shape `[T, T]`, which we don't have here.

Maybe `_masked_softmax` isn't equipped to handle these shapes on CPU or that check is wrong (then we should remove it).

Comment (cpuhrsch) [reply torch/masked/maskedtensor/_ops_refs.py:316]: I think that check you're referencing should be removed if the dimension of the mask matches the dimension of the input. They're saying explicitly `[T, T]` is not supported, but our mask here should have dimension 4.

Comment (cpuhrsch) [reply torch/masked/maskedtensor/_ops_refs.py:316]: Looking at [the CUDA kernel](https://github.com/pytorch/pytorch/blob/b182f0813519d9b153bf7acb4214c8e3f795866e/aten/src/ATen/native/cuda/SoftMax.cu#L964-L978) passing `1` as the mask type indicates a mask of shape `[B, T]` if it's 2-dim.

I'm worried someone might edit the CUDA kernel in the future and will raise an error due to this use site.

Maybe we should introduce a mask_type `2` that indicates the mask is of full dimension (which is our case).

Comment (george-qi) [reply torch/masked/maskedtensor/_ops_refs.py:316]: High level, the current implementation should still work because it’ll just have is_BxT_mask = 0 and then continue throughout normally since `mask.sizes() == input.sizes()`

But I agree it’s not the intended use case of `mask_type=1` and they could possibly add more error checks, so I’m happy with adding another `mask_type = 2`

Will create another PR to precede this one

**Review by cpuhrsch — APPROVED** [review]
Approved, but please address comment before merging.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your PR but it breaks CUDA test_softmax_cuda (__main__.TestBasicsCUDA)' -c ignoresignal

Here are some examples:

* https://hud.pytorch.org/pytorch/pytorch/commit/a4d10342e98b0abb3286a3780617afe108328ac7
* https://hud.pytorch.org/pytorch/pytorch/commit/84a06d71936e61ceeee2abb9c9cb7bf5ee6440dd

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your PR but it breaks CUDA test_softmax_cuda (main.TestBasicsCUDA)' -c ignoredsignal

Here are some examples:

https://hud.pytorch.org/pytorch/pytorch/commit/a4d10342e98b0abb3286a3780617afe108328ac7
https://hud.pytorch.org/pytorch/pytorch/commit/84a06d71936e61ceeee2abb9c9cb7bf5ee6440dd

**Comment by george-qi** [comment]
Fixed the error in the revert - sorry about not catching that due to the CUDA OOM error that was happening. Merging again!
